### PR TITLE
/dev should not be readonly with --readonly flag

### LINKED ISSUE #35041: ReadOnlyRootFs causes /dev tmpfs to be mounted in ro mode
**Description**

Using `ReadOnlyRootFs` causes /dev to be mounted as readonly (though /dev/pts and /dev/shm do not suffer the same fate). This breaks rsyslogd, which wants to create its socket at /dev/log.

I've attempted a few workarounds, including a manual rw tmpfs mount at /dev (which works, but wipes out the devpts mount at /dev/pts) -- I can then bind mount /dev/pts:/dev/pts, but I'm displeased with the security implications of this.

This seems to be the code which is responsible for populating the mount table: https://github.com/moby/moby/blob/a5f9783c930834b8e6035fb0ad9c22fd4bbfc355/daemon/oci_linux.go#L582

**Steps to reproduce the issue:**

```
$ docker run --rm -it --read-only debian:stable bash -c 'mount | grep "on /dev/pts type"'
```

**Describe the results you received:**

By default, /dev is mounted rw:

```
$ docker run --rm -it debian:stable bash -c 'mount | grep "on /dev type"'
tmpfs on /dev type tmpfs (rw,nosuid,mode=755)
```

Using `--read-only`, /dev/pts is still mounted rw:

```
$ docker run --rm -it --read-only debian:stable bash -c 'mount | grep "on /dev/pts type"'
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=666)
```

However, /dev is mounted ro:

```
$ docker run --rm -it --read-only debian:stable bash -c 'mount | grep "on /dev type"'
tmpfs on /dev type tmpfs (ro,mode=755) 
```

When we manually create a rw tmpfs on /dev, we lose the /dev/pts mount entirely: 

```
$ docker run --rm -it --read-only --tmpfs=/dev:rw,dev debian:stable bash -c 'mount | grep "on /dev/pts type"'
**EMPTY OUTPUT**
```

**Describe the results you expected:**

```
$ docker run --rm -it --read-only debian:stable bash -c 'mount | grep "on /dev type"'
tmpfs on /dev type tmpfs (rw,nosuid,mode=755)
```

**or**

```
$ docker run --rm -it --read-only --tmpfs=/dev:rw,dev debian:stable bash -c 'mount | grep "on /dev/pts type"'
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=666)
```

**Output of `docker version`:**

```
$ docker version
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:09:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:09:06 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

### PR DESCRIPTION
**Pull request by rhatdan**
/dev is mounted on a tmpfs inside of a container.  Processes inside of containers
some times need to create devices nodes, or to setup a socket that listens on /dev/log
Allowing these containers to run with the --readonly flag makes sense.  Making a tmpfs
readonly does not add any security to the container, since there is plenty of places
where the container can write tmpfs content.

I have no idea why /dev was excluded.

fixes https://github.com/moby/moby/issues/35041

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @tonistiigi @justincormack @kolyshkin PTAL

**Comment by justincormack** [comment]
There are not plenty of places where there is a writeable tmpfs, currently there is only one that is not yet removeable (`/dev/mqueue` which acts as a tmpfs). Since we added `--ipc=none` support recently we removed the second to last one, so we are close to being able to have totally unwriteable containers. So I am not in favour of always making `/dev` writeable always. I don't mind adding an option to make it writeable, or even the reverse, so long as there is a way to make it not writeable.

**Comment by rhatdan** [comment]
/dev/shm?

**Comment by rhatdan** [comment]
What is the goal of --readonly.  I see it as preventing a process from making changes to the image. Not to where it can not write on any temporary file system.

**Comment by rhatdan** [comment]
```
       --read-only=true|false
          Mount the container's root filesystem as read only.
```
```
docker run --help  | grep root-readonly
      --read-only                   Mount the container's root filesystem as read only
```
Man page and help indicate that the "root" file system is readonly not all file systems inside of the image.

**Comment by justincormack** [comment]
`--ipc=none` removes `/dev/shm`. I think the wording is a bit unclear, and while yes in theory it was the rootfs, which corresponds exactly to the runtime spec config, it always affected some other filesystems. I would be happy if these were affected by other flags instead.

**Comment by rhatdan** [comment]
I guess I don't understand the goal of --read-only flag then.  The way I would describe it, is the flag gives a security feature where a container can not modify the image used for it, and just restarting the container would remove any content created by processes running inside the container, except for content volume mounted into the container.

If your definition is that --read-only means that the processes can not write to anywhere inside of the container except for volumes mounted into the container, then you are failing at that by overriding

"/proc", "/dev/pts", "/dev/mqueue": 

IE Processes would be able to write to /dev/pts, and /dev/mqueue, as well as able to write to /proc.

```
# docker run --read-only -ti fedora mount | grep rw
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime) devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,context="system_u:object_r:container_file_t:s0:c163,c465",gid=5,mode=620,ptmxmode=666)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime,seclabel)
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,context="system_u:object_r:container_file_t:s0:c163,c465",size=65536k)
devpts on /dev/console type devpts (rw,nosuid,noexec,relatime,seclabel,gid=5,mode=620,ptmxmode=000)
```

By either definition --read-only is broken. 

From a security point of view, the security goal is that by default a hacked container can not write a `backdoor` so that if the container was restarted it would be already compromised.

I actually believe all containers in production should run with --read-only rootfs, with /run /var/tmp, /tmp with tmpfs mounted and writing to /dev/*,  Buf if users need to do lots of toggling of flags to make tmpfs work, then --read-only will hardly ever being used, which means we miss an opportunity to make containers more secure.

**Comment by thaJeztah** [comment]
ping @n4ss as well (for entitlements)

**Comment by justincormack** [comment]
You cant actually write stuff to `/dev/pts` as it is a proper pseudo filesystem that only accepts `pty` nodes not arbitrary files.

However, having discovered that you can create a file using `memfd_create` and then use `execveat` to execute it, I hereby give up on not providing writeable tmpfs to containers, so I guess this change is ok...

**Comment by rhatdan** [comment]
@thaJeztah Tests pass who else do I need to get approval from to get this merged?

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
ping @justincormack @kolyshkin PTAL

**Comment by kolyshkin** [comment]
LGTM

I admit I also do not entirely understand why `--read-only` is not specific to container's rootfs only. Was it done as a way to prevent creating/running a new binary?

> However, having discovered that you can create a file using memfd_create and then use execveat to
> execute it, I hereby give up on not providing writeable tmpfs to containers, so I guess this change is ok...

Can seccomp be used to disallow `memfd_create`?

**Comment by kolyshkin** [comment]
Also, `/dev` is (hopefully) mounted with `noexec` so you can't really run a binary from it.

**Comment by kolyshkin** [comment]
Oops, looks like it's not
```
kir@kd:~$ docker run -it alpine mount | grep 'on /dev '
tmpfs on /dev type tmpfs (rw,nosuid,size=65536k,mode=755)
```

Unless there's something we need to run from `/dev`, I think this "no-noexec" should be fixed before merging this PR.

**Comment by n4ss** [comment]
@justincormack would it makes sense to do the following as part a `fs.read-only` entitlement? 
- prevent `memfd_create` syscall
- remove /dev/shm

I get the feeling that the real read-only is going to come from IMA+EVM..

I agree with @kolyshkin, let's make sure /dev has the most restrictive mount option besides `ro`.

**Comment by justincormack** [comment]
well, `memfd_create` is really useful and I have lots of code that uses it...

I did think `/dev` was mounted `noexec` I will have to check if it changed at some point.

**Review by justincormack — APPROVED** [review]

**Comment by justincormack** [comment]
I think we can change `noexec`` in another pr, it is unrelated.

**Comment by thaJeztah** [comment]
I agree we can do the other changes in a follow-up

@n4ss @kolyshkin could one of you make a pull request or open an issue to track that effort?

**Comment by n4ss** [comment]
@thaJeztah https://github.com/moby/moby/issues/35397

**Pull request merged**
